src/Containers/Microsoft.NET.Build.Containers/ImageIndexGenerator.cs

```diff
@@ -0,0 +1,128 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Text.Json;
+using System.Text.Json.Nodes;
+using Microsoft.NET.Build.Containers.Resources;
+using Microsoft.NET.Build.Containers.Tasks;
+
+namespace Microsoft.NET.Build.Containers;
+
+internal struct ImageInfo
+{
+    internal string Config { get; init; }
+    internal string ManifestDigest { get; init; }
+    internal string Manifest { get; init; }
+    internal string ManifestMediaType { get; init; }
+}
+
+internal static class ImageIndexGenerator
+{
+    /// <summary>
+    /// Generates an image index from the given images.
+    /// </summary>
+    /// <param name="imageInfos"></param>
+    /// <returns>Returns json string of image index and image index mediaType.</returns>
+    /// <exception cref="ArgumentException"></exception>
+    /// <exception cref="NotSupportedException"></exception>
+    internal static (string, string) GenerateImageIndex(ImageInfo[] imageInfos)
+    {
+        if (imageInfos.Length == 0)
+        {
+            throw new ArgumentException(string.Format(Strings.ImagesEmpty));
+        }
+
+        string firstManifestMediaType = imageInfos[0].ManifestMediaType;
+
+        if (firstManifestMediaType == SchemaTypes.DockerManifestV2)
+        {
+            return GenerateDockerManifestList(imageInfos, firstManifestMediaType);
+        }
+        else if (firstManifestMediaType == SchemaTypes.OciManifestV1)
+        {
+            return GenerateOciImageIndex(imageInfos, firstManifestMediaType);
+        }
+        else
+        {
+            throw new NotSupportedException(string.Format(Strings.UnsupportedMediaType, firstManifestMediaType));
+        }
+    }
+
+    private static (string, string) GenerateDockerManifestList(ImageInfo[] images, string firstManifestMediaType)
```

Comment:
This function is almost identical to the `GenerateDockerManifestList` one.
Would merging the two either by templating them or by switching the type make sense?
```private static (string, string) GenerateManifestList<TManifest, TManifestList> (ImageInfo[] Images, SchemaType mediaType) { }...
```
also, we can get the firstManifestMediaType in the function (or infer it by the type of the function) so passing it as a parameter looks somewhat redundant.

Reply:
I agree the code repetition looks bad here. The problem here is that `PlatformSpecificManifest` and `ManifestListV2` (similarly in `GenerateOciImageIndex` - `PlatformSpecificOciManifest` and `ImageIndexV1`) are not classes but structs, so they cannot be united by one parent. `PlatformSpecificManifest` and `PlatformSpecificOciManifest` have a slight difference in `annotations` but I don't need this in here, and I could use potentially `ManifestListV2` for both and shorten the code.  I made a choice to be more explicit here and state exactly what struct we are building - with a downside being the code duplication.
However, I don't feel strongly about this choice. I will leave this conversation unresolved for now, to get more feedback.

Reply:
I'd strongly vote for removing code repetition.
One way to unify is to have the struct expose a common interface.
Downside of interface is the boxing/unboxing cost - however both methods doesn't feel like something that will be called in a tight loop - so that would be a very negligable cost

Reply:
+1 on let's get rid of the `firstManifestMediaType` arg - that's an implementation detail of the method

Reply:
Thanks! I refactored media type validation and it looks better now. 
I also tried to use interfaces for the structs
```csharp
internal interface IPlatformSpecificManifest
{
    string mediaType { get; set; }
    long size { get; set; }
    string digest { get; set; }
    PlatformInformation platform { get; set; }
}

internal interface IImageIndex
{
    int schemaVersion { get; set; }
    string mediaType { get; set; }
    IPlatformSpecificManifest[] manifests { get; set; }
}
```
but the problem is I need to change structs 
```diff
- public record struct ManifestListV2(int schemaVersion, string mediaType, PlatformSpecificManifest[] manifests);
+ public record struct ManifestListV2(int schemaVersion, string mediaType, IPlatformSpecificManifest[] manifests) : IImageIndex;
```
same for `ImageIndexV1` struct. Because they are used in other places, probably for performance reasons, the boxing/unboxing will affect it negatively.

There is also a solution to only use `ManifestListV2`(or `ImageIndexV1`) in `ImageIndexGenerator`, that would reduce code duplication. I'm going with this route. And if in the future we would need extra fields like "annotations" in `ImageIndexV1`, I'll just change the code (not likely to happen)